RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc

```diff
@@ -0,0 +1,611 @@
+    double delta,
+    unsigned trackstersSize,
+    std::vector<std::vector<unsigned>> &resultCollection,
+    bool useMask = false) {
+  // Finds tracksters in tracksterTiles within an eta-phi window
+  // (given by delta) of the objects (track/trackster) in the seedingCollection.
+  // Element i in resultCollection is the vector of trackster
+  // indices found close to the i-th object in the seedingCollection.
+  // If specified, Tracksters are masked once found as close to an object.
+  std::vector<int> mask(trackstersSize, 0);
+
+  for (auto &i : seedingCollection) {
+    auto seed_eta = i.first.Eta();
+    auto seed_phi = i.first.Phi();
+    unsigned seedId = i.second;
+
+    if (seed_eta > 0.) {
+      // look in the forward tile
+      const TICLLayerTile &tile = tracksterTiles[1];
+      double eta_min = std::max(seed_eta - delta, (double)TileConstants::minEta);
+      double eta_max = std::min(seed_eta + delta, (double)TileConstants::maxEta);
+
+      std::array<int, 4> search_box = tile.searchBoxEtaPhi(eta_min, eta_max, seed_phi - delta, seed_phi + delta);
+      if (search_box[2] > search_box[3])
+        search_box[3] += TileConstants::nPhiBins;
+
+      for (int eta_i = search_box[0]; eta_i <= search_box[1]; ++eta_i) {
+        for (int phi_i = search_box[2]; phi_i <= search_box[3]; ++phi_i) {
+          const auto &in_box = tile[tile.globalBin(eta_i, (phi_i % TileConstants::nPhiBins))];
+          for (const unsigned t_i : in_box) {
+            if (!mask[t_i]) {
+              resultCollection[seedId].push_back(t_i);
+              if (useMask)
+                mask[t_i] = 1;
+            }
+          }
+        }
+      }
+    }  // forward
+
+    else if (seed_eta < 0.) {
+      // backward tile
+      const TICLLayerTile &tile = tracksterTiles[0];
+      double eta_min = std::max(abs(seed_eta) - delta, (double)TileConstants::minEta);
+      double eta_max = std::min(abs(seed_eta) + delta, (double)TileConstants::maxEta);
+
+      std::array<int, 4> search_box = tile.searchBoxEtaPhi(eta_min, eta_max, seed_phi - delta, seed_phi + delta);
+      if (search_box[2] > search_box[3])
+        search_box[3] += TileConstants::nPhiBins;
+
+      for (int eta_i = search_box[0]; eta_i <= search_box[1]; ++eta_i) {
+        for (int phi_i = search_box[2]; phi_i <= search_box[3]; ++phi_i) {
+          const auto &in_box = tile[tile.globalBin(eta_i, (phi_i % TileConstants::nPhiBins))];
+          for (const unsigned t_i : in_box) {
+            if (!mask[t_i]) {
+              resultCollection[seedId].push_back(t_i);
+              if (useMask)
+                mask[t_i] = 1;
+            }
+          }
+        }
+      }
+    }  // backward
+  }    // seeding collection loop
+}
+
+void LinkingAlgoByDirectionGeometric::dumpLinksFound(std::vector<std::vector<unsigned>> &resultCollection,
+                                                     const char *label) const {
+  if (!(LinkingAlgoBase::algo_verbosity_ > LinkingAlgoBase::Advanced))
```

Comment:
you could also wrap the contents of this function in `#ifdef EDM_ML_DEBUG` so that it's compiled if and only if debug is enabled.
it may be confusing for non-TICL people in the future to navigate multiple parallel levels of message logging.

Reply:
now the function implementation is not defined if `EDM_ML_DEBUG` is not defined. In fact I'm confused why scram does not crash at linking due to a missing implementation.

I thought something like this might be desirable here:
```C++
void LinkingAlgoByDirectionGeometric::dumpLinksFound(std::vector<std::vector<unsigned>> &resultCollection,
                                                     const char *label) const {
#ifdef EDM_ML_DEBUG
...
#endif
}
```